Update JEI Multiblocks to show block counts and allow bookmarking

warjort opened the pull request:
**What:**
This shows the number of blocks required to build the currently shown pattern in the JEI multiblock view and allows bookmarking of those blocks.

**How solved:**
Added 20 slots at the bottom of the screen. 
These are filled from the block drops of the multiblock in the dummy world, which was sort of already being collected (just not the amounts).
The new slots are part of an IGuiItemStackGroup to enable all the extra JEI features like bookmarking and r/u behaviour. 
The left/right buttons to switch between multiblock variants have been moved to the right hand side below the layer switch button.

![2021-02-09_00 26 12](https://user-images.githubusercontent.com/275258/107298841-0f441900-6a6e-11eb-9a4c-e62a10bad835.png)

**Outcome:**
Closes #1393 
and does part 1 of #1211

**Possible compatibility issue:**
I made it 20 slots because I know gregicality has some complicated multiblocks.

LAGIdiot commented:
Thanks for you hard work. Please update your opening comment with screenshots as you are changing GUI.

Regarding #1393 it provides requested functionality, probably not in way that @ALongStringOfNumbers expected but it should be usable right?

For #1211 It would be better to specify what part were implemented so we can update that issue after this one is closed.

I have quite few points regarding appearance of this update. 
![2021-02-08_18 28 12](https://user-images.githubusercontent.com/3093101/107258802-8496f600-6a3c-11eb-9ac6-d2c9cb71144f.png)
I think we should render parts little bit closer to bottom and align it to center of window.
Scene should be rendered from top of GUI and possibly leave some space between.
Buttons with arrows should not touch parts area.

![2021-02-08_18 28 18](https://user-images.githubusercontent.com/3093101/107258807-85c82300-6a3c-11eb-89ff-6c9d5f62cdfe.png)
Parts should have same order for all variants and multiblock types. Possibly starting with Controller, followed by active blocks then passive blocks.

![2021-02-08_18 28 48](https://user-images.githubusercontent.com/3093101/107258809-8660b980-6a3c-11eb-994f-8238cfec8a8e.png)
Sometimes hovering over some buttons highlight parts slots, and sometimes they grey out.

![2021-02-08_18 29 09](https://user-images.githubusercontent.com/3093101/107258811-8660b980-6a3c-11eb-85b7-40199df748c2.png)
Hovering over bottom 1/3 of part does not show tooltip.

warjort commented:
For reference this is what I see with the auto gui scale:
![2021-02-08_13 14 08](https://user-images.githubusercontent.com/275258/107265656-6df09f00-6a3c-11eb-88c9-8056163a65d9.png)

Just so you understand why I crammed it in. There isn't really any space at the top or bottom of the GUI at this scale.

For centering, the left hand side of these slots is already at x=1, I think the only way to make it look centred is to reduce the number of slots to 18 and move the whole thing to the right.

I haven't done anything with tool tips. I will investigate what magic incantation I did to do to make JEI? do this properly. :-)
The only thing I am drawing is the background texture of the slot.
For me the tooltip only seems to work in the top half of the slot.

On the ordering, the only thing I have access to is the BlockInfo which is a position, IBlockState and a TileEntity.
In the original code, it had already gone through a HashMap by the time it calculates the block drops so the order is "random".

Given the limited info available, it could be ordered something like:
* TileEntity is a controller block
* other tile entities
* other blocks
and I could maybe include the number of blocks and the block id to try to keep ordering the same across variants?
Let me know if this is ok, or you have a different idea.

DStrand1 commented:
I think it would be worthwhile to mention some details regarding some of the ingredients too. For example, the EBF requires a minimum of 10 Heat-Proof casings to form, and fluid hatches are optional for it to form.

warjort commented:
> I think it would be worthwhile to mention some details regarding some of the multiblock parts too. For example, the EBF requires a minimum of 10 Heat-Proof casings to form, and fluid hatches are optional for it to form. These could either be appended to the multiblock part tooltip, or replace it in some instances

That information is not available at this part of the code, if it is available at all anywhere.
All that is available are the blocks in the scene/variant you are viewing in the GUI.
The easiest way to provide this extra information atm would be to add something to the default tooltip of the controller blocks.

LAGIdiot commented:
> I think it would be worthwhile to mention some details regarding some of the multiblock parts too. For example, the EBF requires a minimum of 10 Heat-Proof casings to form, and fluid hatches are optional for it to form. These could either be appended to the multiblock part tooltip, or replace it in some instances

This all is mentioned in #1211 and I think we can come up with much more what could be useful.

ALongStringOfNumbers commented:
I am not sure how hard it would be, but to attempt to save space on the gui, you could reduce the number of rows to 1 and then use arrows to show more items. 

An example would be: The multiblock has 18 items, the first 9 get rendered in the slots and you have a `>` button (Similar to how the EBF has one for displaying different coils) that will render the next 9 items in the slots. And then you can go back to the first nine items with a `<` button.

Of course, these arrows could be somewhat big and might require reducing the number of slots show

warjort commented:
> I will investigate what magic incantation I need to do to make JEI? do this properly. :-)

I think I understand why this is not working. The bottom half of the top row of slots is outside the background we have to told JEI to use.

To make this work for the double row of slots, the background would need to be extended by 30 "pixels". Does that sound ok?

warjort commented:
For the space saving, could I move the left and right arrows for changing scenes to be together under the LA button for changing layers? That would free up space at the bottom for the parts slots.

LAGIdiot commented:
> For the space saving, could I move the left and right arrows for changing scenes to be together under the LA button for changing layers? That would free up space at the bottom for the parts slots.

Feel free

warjort commented:
> Sometimes hovering over some buttons highlight parts slots, and sometimes they grey out.

This looks like a bug in minecraft? 

The highlighting of the buttons is sometimes leaking into the draw operations for the parts background texture.
I don't exactly understand the circumstances to always reproduce it and it is not always the last button drawn, but it looks like I have fixed this by changing the code so the buttons are drawn last.

Very weird.

warjort commented:
I think I have done all the things you asked for?

The left/right buttons have been moved to make space for the slots so they now fit on the JEI background properly which fixes the tooltip issue. I didn't change the size of the background so it will now show properly in the auto gui scale.
See the new image in the top post.

I moved the slots to start at x=0 and they look less off centre now the positions of the buttons are not pointing it out.

I have sorted the parts as I suggested above. This was more complicated than expected as the data only becomes available in phases and so has to be collected in multiple steps. Not the most elegant code I ever wrote. :-)

I worked around the button highlight issue by changing the drawing order.

LAGIdiot commented:
I see what you did regarding my request and it seems like you did everything that is possible without rewrite. And I understand that other things are not possible without it.

One other thing that needs to be done is move render scene above parts slots. So rendered multiblock can be hidden behind it.

warjort commented:
I changed the height of the rendered scene so it doesn't try to draw under the part slots.

However, the line I changed looked like a typo. I guess it didn't matter since the area is pretty much square?
````
-int sceneHeight = recipeWidth;
+int sceneHeight = recipeHeight-36;
````

LAGIdiot commented:
> I changed the height of the rendered scene so it doesn't try to draw under the part slots.
> 
> However, the line I changed looked like a typo. I guess it didn't matter since the area is pretty much square?
> 
> ```
> -int sceneHeight = recipeWidth;
> +int sceneHeight = recipeHeight-36;
> ```

I guess it was a typo

LAGIdiot reviewed (approved):
Thanks for finishing this up.
